File: sdks/go/e2e/runner/json_runner.go
package runner

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/compforge/quality-harness/sdks/go/e2e/core"
)

// JSONRunner sends HTTP requests and parses JSON responses into Outcome.
type JSONRunner struct {
	config *core.E2EConfig
	client *http.Client
}

// NewJSONRunner creates a runner for standard JSON REST/RPC APIs.
func NewJSONRunner(config *core.E2EConfig) *JSONRunner {
	timeout := time.Duration(config.Runtime.HTTPTimeoutS) * time.Second
	if timeout <= 0 {
		timeout = 120 * time.Second
	}
	return &JSONRunner{
		config: config,
		client: &http.Client{
			Timeout: timeout,
		},
	}
}

func (r *JSONRunner) Trigger(ctx context.Context, req Request) (*Outcome, error) {
	var bodyReader io.Reader
	if req.Body != nil {
		data, err := json.Marshal(req.Body)
		if err != nil {
			return nil, fmt.Errorf("marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(data)
	}

	method := req.Method
	if method == "" {
		method = http.MethodPost
	}

	url := r.config.Service.BaseURL + req.Path
	httpReq, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}

	// Set headers
	httpReq.Header.Set("Content-Type", "application/json")
	for k, v := range r.config.Service.Headers {
		httpReq.Header.Set(k, v)
	}
	for k, v := range req.Headers {
		httpReq.Header.Set(k, v)
	}
	for _, key := range req.ExcludeHeaders {
		httpReq.Header.Del(key)
	}

	// Set query params
	if len(req.Query) > 0 {
		q := httpReq.URL.Query()
		for k, v := range req.Query {
			q.Set(k, v)
		}
		httpReq.URL.RawQuery = q.Encode()
	}

	start := time.Now()
	resp, err := r.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()
	durationMS := time.Since(start).Milliseconds()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response body: %w", err)
	}

	// Parse response headers
	headers := make(map[string]string, len(resp.Header))
	for k := range resp.Header {
		headers[k] = resp.Header.Get(k)
	}

	// Parse JSON body if content-type is JSON
	var body json.RawMessage
	ct := resp.Header.Get("Content-Type")
	if len(raw) > 0 && (ct == "" || bytes.Contains([]byte(ct), []byte("json"))) {
		if json.Valid(raw) {
			body = raw
		}
	}

	return &Outcome{
		StatusCode: resp.StatusCode,
		Body:       body,
		Headers:    headers,
		DurationMS: durationMS,
		Metadata:   map[string]any{},
		Raw:        raw,
	}, nil
}

File: sdks/python/e2e_harness/runner/json_runner.py
"""JSONRunner — synchronous JSON API runner.

Sends HTTP requests and parses JSON responses into Outcome.
"""

from __future__ import annotations

import time
from typing import Any

import httpx

from e2e_harness.core.config import E2EConfig
from e2e_harness.runner.base import BaseRunner, Outcome, Request
from e2e_harness.runner.headers import build_headers


class JSONRunner(BaseRunner):
    """Runner for standard JSON REST/RPC APIs."""

    def __init__(self, config: E2EConfig, *, client: httpx.Client | None = None):
        self._env = config
        self._client = client or httpx.Client(
            base_url=config.service.base_url,
            timeout=httpx.Timeout(config.runtime.http_timeout_s),
        )

    def trigger(self, request: Request) -> Outcome:
        headers = build_headers(
            self._env, extra=request.headers, exclude=request.exclude_headers
        )

        start = time.monotonic()
        resp = self._client.request(
            method=request.method,
            url=request.path,
            json=request.body,
            headers=headers,
            params=request.query or None,
        )
        duration_ms = int((time.monotonic() - start) * 1000)

        body = self._parse_body(resp)

        return Outcome(
            status_code=resp.status_code,
            body=body,
            headers=dict(resp.headers),
            duration_ms=duration_ms,
            metadata={},
            raw=resp.content,
        )

    @staticmethod
    def _parse_body(resp: httpx.Response) -> dict[str, Any] | None:
        ct = resp.headers.get("content-type", "")
        if "json" not in ct:
            return None
        try:
            return resp.json()
        except Exception:
            return None

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> JSONRunner:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()
